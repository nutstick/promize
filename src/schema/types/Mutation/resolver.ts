import { IResolver } from '../index';
// tslint:disable-next-line:no-var-requires
const m = require('casual');

const resolver: IResolver<any, any> = {
  Mutation: {
    async editPrice(_, { product, price }, { database }) {
      await database.Product.update({ _id: product }, {
        $set: {
          price,
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async addPicture(_, { product, pictures }, { database }) {
      await database.Product.update({ _id: product }, {
        $push: {
          picture: pictures,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async addHashtag(_, { product, hashtags }, { database }) {
      await database.Product.update({ _id: product }, {
        $push: {
          hashtag: hashtags,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async addSize(_, { product, size }, { database }) {
      await database.Product.update({ _id: product }, {
        $push: {
          sizes: size,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async addColor(_, { product, color }, { database }) {
      await database.Product.update({ _id: product }, {
        $push: {
          colors: color,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async removePicture(_, { product, pictures }, { database }) {
      await database.Product.update({ _id: product }, {
        $pull: {
          picture: pictures,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async removeHashtag(_, { product, hashtags }, { database }) {
      await database.Product.update({ _id: product }, {
        $pull: {
          hashtag: hashtags,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async removeColor(_, { product, color }, { database }) {
      await database.Product.update({ _id: product }, {
        $push: {
          colors: color,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async removeSize(_, { product, size }, { database }) {
      await database.Product.update({ _id: product }, {
        $push: {
          sizes: size,
        },
        $set: {
          updateAt: Date.now(),
        },
      });
      return await database.Product.findOne({ _id: product });
    },
    async createProduct(_, { input }, { database }) {
      return await database.Product.insert({
        ...input,
        promotion_start: input.promotionStart,
        promotion_end: input.promotionEnd,
      });
    },
  },
};

export default resolver;
