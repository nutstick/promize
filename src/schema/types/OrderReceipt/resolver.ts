import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
  OrderReceiptEdges: {
  },
  OrderReceipt: {
    async deliverAddress({ buyer, deliver_address }, _, { database }) {
      const creator_data = await database.User.findOne({ _id: buyer });
      for (const i of creator_data.addresses) {
        if (i._id.toString() === deliver_address) {
          return i;
        }
      }
    },
    trackingId({ tracking_id }) {
      return tracking_id;
    },
    async paymentMethod({ buyer, payment_method }, _, { database }) {
      const creator_data = await database.User.findOne({ _id: buyer });
      for (const i of creator_data.payment_methods) {
        if (i._id.toString() === payment_method) {
          return i;
        }
      }
    },
    async color({ product, color }, _, { database }) {
      const product_data = await database.Product.findOne({ _id: product });
      for (const i of product_data.colors) {
        if (i._id.toString() === color) {
          return i;
        }
      }
    },
    async size({ product, size }, _, { database }) {
      const product_data = await database.Product.findOne({ _id: product });
      for (const i of product_data.sizes) {
        if (i._id.toString() === size) {
          return i;
        }
      }
    },
    status({ payment_completed, product_delivered, product_received }) {
      return payment_completed ? 'PAID' : product_delivered ? 'DELIVERED' : product_received ?
        'RECEIVED' : 'CREATED';
    },
    paymentCompletedAt({ payment_completed_at }) {
      return payment_completed_at;
    },
    productDeliveredAt({ product_delivered_at }) {
      return product_delivered_at;
    },
    productReceivedAt({ product_received_at }) {
      return product_received_at;
    },
    async product({ product }, _, { database }) {
      return await database.Product.findOne({ _id: product });
    },
    async creator({ buyer }, _, { database }) {
      return await database.User.findOne({ _id: buyer });
    },
  },
};

export default resolver;
