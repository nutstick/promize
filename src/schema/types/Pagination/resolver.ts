import { GraphQLError } from 'graphql';
import { toObjectID } from 'iridium';
import { Cursor as MongoCursor } from 'mongodb';

export type Base64String = string;

export function base64(i: string): Base64String {
  return Buffer.from(i, 'utf8').toString('base64');
}

export function unbase64(i: Base64String): string {
  return Buffer.from(i, 'base64').toString('utf8');
}

type Cursor = string;

interface SortField {
  sort: string;
  direction: 'ASC' | 'DESC';
}

export interface PaginationOption {
  name: string;
  first?: number;
  after?: Cursor;
  last?: number;
  before?: Cursor;
  orderBy?: SortField[];
}

export function toCursor(node: any, name: string, orderBy: SortField[]) {
  return base64(JSON.stringify(
    orderBy.reduce((prev, order) => ({
      ...prev,
      [order.sort]: node[order.sort],
    }), { name }),
  ));
}

export async function pagination<T = any>(cursor: MongoCursor<T>, options: PaginationOption) {
  const { after, first, before, last } = options;

  // Add default sort by _id
  const orderBy: SortField[] = options.orderBy ? options.orderBy.concat([{
    sort: '_id', direction: 'ASC',
  }]) : [{ sort: '_id', direction: 'ASC' }];

  // Parse orderBy to mongodb sort format
  const sort: any = orderBy ? orderBy.reduce((prev, order) => ({
    ...prev,
    [order.sort]: (order.direction === 'ASC' ? 1 : -1),
  }), {}) : {};

  if (after) {
    try {
      const { name, _id, ...cursorData } = JSON.parse(unbase64(after));

      if (name !== options.name) {
        throw new GraphQLError(`Type of 'after' is not a ${name}`);
      }
      // FIXME: @types/MongoDB Cursor.min(number)
      cursor = (cursor as any).min({
        ...cursorData,
        ..._id && {
          _id: toObjectID(_id),
        },
      });
    } catch (err) {
      if (err instanceof GraphQLError) {
        throw err;
      } else {
        throw new Error(`Can't parse cursor on field 'after' make use you use Cursor type`);
      }
    }
  }
  if (first) {
    cursor = cursor.limit(first + 1);
  }
  if (before) {
    try {
      const { name, _id, ...cursorData } = JSON.parse(unbase64(before));
      if (name !== options.name) {
        throw new GraphQLError(`Type of 'before' is not ${options.name}`);
      }
      // FIXME: @types/MongoDB Cursor.min(number)
      cursor = (cursor as any).max({
        ...cursorData,
        ..._id && {
          _id: toObjectID(_id),
        },
      });
    } catch (err) {
      if (err instanceof GraphQLError) {
        throw err;
      } else {
        throw new Error(`Can't parse cursor on field 'before' make use you use Cursor type`);
      }
    }
  }
  if (last) {
    cursor = cursor.limit(last + 1);
  }

  cursor = cursor.sort(sort);

  const totalCount = await cursor.count();
  const results = await cursor.toArray();

  if (results.length > (first || last)) {
    const endCursor = toCursor(results[results.length - 1], options.name, orderBy);

    return {
      totalCount,
      edges: results.slice(0, -1).map((node) => ({
        node,
        cursor: toCursor(node, options.name, orderBy),
      })),
      pageInfo: {
        endCursor,
        hasNextPage: true,
      },
    };
  }

  return {
    totalCount,
    edges: results.map((node) => ({
      node,
      cursor: toCursor(node, options.name, orderBy),
    })),
    pageInfo: {
      endCursor: null,
      hasNextPage: false,
    },
  };
}
