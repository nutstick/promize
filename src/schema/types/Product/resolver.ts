import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
  ProductEdges: {
  },
  Product: {
    promotionStart({ promotion_start }) {
      return promotion_start;
    },
    promotionEnd({ promotion_end }) {
      return promotion_end;
    },
    originalPrice({ orginal_price }) {
      return orginal_price;
    },
    async owner({ owner }, _, { database }) {
      return await database.User.findOne(owner);
    },
  },
};

export default resolver;
