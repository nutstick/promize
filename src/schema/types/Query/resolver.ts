import * as BluebirdPromise from 'bluebird';
import * as fs from 'fs';
import { join } from 'path';
import { locales } from '../../../config';
import { IResolver } from '../index';

// A folder with messages
// In development, source dir will be used
const MESSAGES_DIR = process.env.MESSAGES_DIR || join(__dirname, './messages');

const readFile = BluebirdPromise.promisify(fs.readFile);

const resolver: IResolver<any, any> = {
  Query: {
    helloworld() {
      return 'Hello Word';
    },
    async me(_, __, { database, user }) {
      if (user && user._id) {
        return await database.User.findOne({ _id: user._id });
      }
      return null;
    },
    async intl(_, { locale }) {
      if (!locales.includes(locale)) {
        throw new Error(`Locale '${locale}' not supported`);
      }

      let localeData;
      try {
        localeData = await readFile(join(MESSAGES_DIR, `${locale}.json`));
      } catch (err) {
        if (err.code === 'ENOENT') {
          throw new Error(`Locale '${locale}' not found`);
        }
      }

      return JSON.parse(localeData);
    },
  },
};

export default resolver;
