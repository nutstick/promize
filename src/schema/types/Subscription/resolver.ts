import { withFilter } from 'graphql-subscriptions';
import { IMessageDocument } from '../../models/Message';
import { ISubsciption } from '../index';
import { MESSAGE_ADDED, pubsub } from '../pubsub';

const resolver: ISubsciption = {
  Subscription: {
    messageAdded: {
      resolve(message) {
        if (message._id) {
          return message;
        }
        return null;
      },
      subscribe: withFilter(
        () => pubsub.asyncIterator(MESSAGE_ADDED),
        (message: IMessageDocument, args) => {
          return message && message.traderoom === args.traderoom;
        },
      ),
    },
  },
};

export default resolver;
