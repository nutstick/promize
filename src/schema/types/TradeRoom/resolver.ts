import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
};

export default resolver;
