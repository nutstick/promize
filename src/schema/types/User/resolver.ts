import { IResolver } from '../index';

const resolver: IResolver<any, any> = {
  User: {
  },
};

export default resolver;
