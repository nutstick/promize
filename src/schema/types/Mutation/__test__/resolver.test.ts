import { graphql } from 'graphql';
import { Schema } from '../../../';
import { mongodb } from '../../../../config';
import { Database } from '../../../models';
import resolver from '../resolver';

const database = new Database({
  ...mongodb,
  database: 'test',

});

let user;

beforeAll(async () => {
  await database.connect();
  await database.connection.dropDatabase();

  // Fixed new Date
  const FIXED_DATE = new Date(Date.UTC(2017, 7, 9, 8));
  (global as any).Date = jest.fn((...input) => FIXED_DATE);

  // Init test database
  user = await database.User.create({
    _id: '5a10494ee67657f236e6b0a2',
    first_name: 'Ammarin',
    last_name: 'Jetthakun',
    tel_number: '0123456789',
    account: {
      email: 'ammarinjtk@gmail.com',
      password: 'donttell',
    },
    avatar: 'someurl',
  });

  // await database.Product.create({
  //   name: 'Zara',
  //   description: 'Zara clothes',
  //   original_price: 1000,
  //   type: 'BuyNowProduct',
  //   price: 800,
  //   picture: ['https://th-live-02.slatic.net/p/7/hequ-1483111676-123106' +
  //     '5-c566b543a82cfe5a0e279dbf161bd13e-catalog_233.jpg'],
  //   hashtag: ['uniqlo'],
  //   colors: ['red'],
  //   sizes: ['S'],
  //   promotion_start: new Date(2017, 13, 1, 12),
  //   promotion_end: new Date(2017, 14, 1, 19),
  //   owner: user._id,
  //   createAt: new Date(2017, 13, 1, 8),
  //   updateAt: new Date(2017, 13, 1, 8),
  // });
});

afterAll(() => database.close());

it('Mutation createProduct should insert new product into mongodb', async () => {
  await resolver.Mutation.createProduct({}, {
    input: {
      _id: '585b11e7adb8b5f2d655da01',
      name: 'a',
      type: 'Product',
      pictures: ['https://th-live-02.slatic.net/p/7/hequ-1483111676-123106' +
        '5-c566b543a82cfe5a0e279dbf161bd13e-catalog_233.jpg'],
      hashtags: ['a', 'b', 'c'],
      colors: ['red'],
      size: ['S'],
      promotionStart: new Date(),
      promotionEnd: new Date(),
      owner: user._id,
    },
  }, { database });

  // Number of product in database should be 1.
  const count = await database.Product.find().count();
  expect(count).toMatchSnapshot();
  // Product should create in database.
  const products = await database.Product.find().toArray();
  expect(products).toMatchSnapshot();
  // Should be able to find a product that has been created.
  const product = await database.Product.findOne({ name: 'a' });
  expect(product).toMatchSnapshot();
});

// it('Mutation editPrice should change price of product', async () => {
//   const product = await database.Product.findOne({
//     name: 'Zara',
//   });
//   await resolver.Mutation.editPrice({
//     product: product._id,
//     price: 100,
//   });

//   const expect_product = await database.Product.findOne({
//     name: 'Zara',
//   });
//   expect(expect_product.price)
//     .toEqual(100);
// });
