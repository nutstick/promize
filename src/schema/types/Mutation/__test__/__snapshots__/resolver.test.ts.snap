// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation createProduct should insert new product into mongodb 1`] = `[Function]`;

exports[`Mutation createProduct should insert new product into mongodb 2`] = `
Array [
  Object {
    "_id": "585b11e7adb8b5f2d655da01",
    "colors": Array [
      "red",
    ],
    "createAt": 2017-06-12T21:41:20.000Z,
    "hashtags": Array [
      "a",
      "b",
      "c",
    ],
    "name": "a",
    "owner": "5a10494ee67657f236e6b0a2",
    "pictures": Array [
      "https://th-live-02.slatic.net/p/7/hequ-1483111676-1231065-c566b543a82cfe5a0e279dbf161bd13e-catalog_233.jpg",
    ],
    "promotionEnd": 2017-06-12T21:41:20.000Z,
    "promotionStart": 2017-06-12T21:41:20.000Z,
    "promotion_end": 2017-06-12T21:41:20.000Z,
    "promotion_start": 2017-06-12T21:41:20.000Z,
    "size": Array [
      "S",
    ],
    "type": "Product",
    "updateAt": 2017-06-12T21:41:20.000Z,
  },
]
`;

exports[`Mutation createProduct should insert new product into mongodb 3`] = `
Object {
  "_id": "585b11e7adb8b5f2d655da01",
  "colors": Array [
    "red",
  ],
  "createAt": 2017-06-12T21:41:20.000Z,
  "hashtags": Array [
    "a",
    "b",
    "c",
  ],
  "name": "a",
  "owner": "5a10494ee67657f236e6b0a2",
  "pictures": Array [
    "https://th-live-02.slatic.net/p/7/hequ-1483111676-1231065-c566b543a82cfe5a0e279dbf161bd13e-catalog_233.jpg",
  ],
  "promotionEnd": 2017-06-12T21:41:20.000Z,
  "promotionStart": 2017-06-12T21:41:20.000Z,
  "promotion_end": 2017-06-12T21:41:20.000Z,
  "promotion_start": 2017-06-12T21:41:20.000Z,
  "size": Array [
    "S",
  ],
  "type": "Product",
  "updateAt": 2017-06-12T21:41:20.000Z,
}
`;
